client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx

```diff
@@ -0,0 +1,217 @@
+import { recordTracksEvent } from '@automattic/calypso-analytics';
+import { FEATURE_SET_PRIMARY_CUSTOM_DOMAIN } from '@automattic/calypso-products';
+import { Card } from '@automattic/components';
+import { useTranslate } from 'i18n-calypso';
+import { useState, ChangeEvent, useEffect, FormEvent } from 'react';
+import CardHeading from 'calypso/components/card-heading';
+import FormButton from 'calypso/components/forms/form-button';
+import FormFieldset from 'calypso/components/forms/form-fieldset';
+import FormSelect from 'calypso/components/forms/form-select';
+import InlineSupportLink from 'calypso/components/inline-support-link';
+import { useSelector } from 'calypso/state';
+import { NON_PRIMARY_DOMAINS_TO_FREE_USERS } from 'calypso/state/current-user/constants';
+import { currentUserHasFlag } from 'calypso/state/current-user/selectors';
+import type { DomainData, SiteDetails } from '@automattic/data-stores';
+
+import './style.scss';
+
+type PrimaryDomainSelectorProps = {
+	domains: undefined | DomainData[];
+	site: undefined | null | SiteDetails;
+	onSetPrimaryDomain: ( domain: string, onComplete: () => void, type: string ) => void;
+};
+
+const PrimaryDomainSelector = ( {
+	domains,
+	site,
+	onSetPrimaryDomain,
+}: PrimaryDomainSelectorProps ) => {
+	const [ selectedDomain, setSelectedDomain ] = useState< undefined | string >( undefined );
+	const [ primaryDomain, setPrimaryDomain ] = useState< undefined | string >( undefined );
+	const [ isSettingPrimaryDomain, setIsSettingPrimaryDomain ] = useState< boolean >( false );
+
+	const translate = useTranslate();
+
+	const primaryWithPlanOnly = useSelector( ( state ) =>
+		currentUserHasFlag( state, NON_PRIMARY_DOMAINS_TO_FREE_USERS )
+	);
+
+	useEffect( () => {
+		if ( domains?.length ) {
+			const primary = domains.find( ( domain ) => {
+				return domain.primary_domain;
+			} );
+			if ( primary ) {
+				setPrimaryDomain( primary.domain );
+			}
+		}
+	}, [ domains ] );
+
+	if ( ! domains || ! site ) {
+		return null;
+	}
+
+	const isOnFreePlan = site?.plan?.is_free ?? false;
+	const canUserSetPrimaryDomainOnThisSite = ! ( primaryWithPlanOnly && isOnFreePlan );
+
+	const shouldUpgradeToMakeDomainPrimary = ( domain: DomainData ): boolean => {
+		return (
+			! canUserSetPrimaryDomainOnThisSite &&
+			( domain.type === 'registered' || domain.type === 'mapping' ) &&
+			! domain.current_user_can_create_site_from_domain_only &&
+			! domain.wpcom_domain &&
+			! domain.is_wpcom_staging_domain &&
+			( site?.plan?.features?.active.includes( FEATURE_SET_PRIMARY_CUSTOM_DOMAIN ) ?? false )
+		);
+	};
+
+	let validPrimaryDomains = domains.filter( ( domain ) => {
+		return (
+			domain.can_set_as_primary &&
+			! domain.aftermarket_auction &&
+			domain.points_to_wpcom &&
+			! shouldUpgradeToMakeDomainPrimary( domain ) &&
+			! domain.primary_domain
+		);
+	} );
+
+	const hasWpcomStagingDomain = validPrimaryDomains.find(
```

Comment:
```suggestion
	const hasWpcomStagingDomain = domains.find(
```

I think we need to check among all domains and not only the ones that can be set as primary 🤔 otherwise this can happen:
<img width="1254" alt="Screenshot 2024-03-18 at 20 12 33" src="https://github.com/Automattic/wp-calypso/assets/5324818/60c8c92f-a914-47f5-907e-637606792faf">
 
The `.wordpress.com` subdomain is not shown in the list, but it's one of the domains that's returned for the site. In this case, the `.wpcomstaging.com` subdomain is not among `validPrimaryDomains` because it's currently the primary domain. I believe this should fix the issue @Aurorum reported too.

Reply:
Good catch!